operator/controllers/resources/teleport_reconciler.go

```diff
@@ -0,0 +1,158 @@
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	// If err is nil, we found the resource. If err != nil (and we did return), then the error was `NotFound`
+	exists := err == nil
+
+	newOwnershipCondition, isOwned := checkOwnership(existingResource)
+	meta.SetStatusCondition(k8sResource.StatusConditions(), newOwnershipCondition)
+	if !isOwned {
+		return trace.NewAggregate(
+			trace.AlreadyExists("unowned resource '%s' already exists", existingResource.GetName()),
+			r.Status().Update(ctx, k8sResource),
+		)
+	}
+
+	teleportResource.SetOrigin(types.OriginKubernetes)
+
+	if !exists {
+		err = r.resourceClient.Create(ctx, teleportResource)
+	} else {
+		/* TODO: handle modifier logic like CreatedBy for users,
+		we can add mutate logic, diffing could also happen here */
+		err = r.resourceClient.Update(ctx, teleportResource)
+	}
+	// If an error happens we want to put it in status.conditions before returning.
+	newReconciliationCondition := getReconciliationConditionFromError(err)
+	meta.SetStatusCondition(k8sResource.StatusConditions(), newReconciliationCondition)
+	if err != nil {
+		return trace.NewAggregate(err, r.Status().Update(ctx, k8sResource))
+	}
+
+	// We update the status conditions on exit
+	return trace.Wrap(r.Status().Update(ctx, k8sResource))
+}
+
+// Delete is the TeleportResourceReconciler of the ResourceBaseReconciler DeleteExertal
+func (r TeleportResourceReconciler[T, K]) Delete(ctx context.Context, obj kclient.Object) error {
+	return r.resourceClient.Delete(ctx, obj.GetName())
+}
+
+// Reconcile allows the TeleportResourceReconciler to implement the reconcile.Reconciler interface
+func (r TeleportResourceReconciler[T, K]) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	kubeResource, err := newKubeResource[T, K]()
+	if err != nil {
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+	return r.Do(ctx, req, kubeResource)
+}
+
+// SetupWithManager have a controllerruntime.Manager run the TeleportResourceReconciler
+func (r TeleportResourceReconciler[T, K]) SetupWithManager(mgr ctrl.Manager) error {
+	kubeResource, err := newKubeResource[T, K]()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return ctrl.NewControllerManagedBy(mgr).For(kubeResource).Complete(r)
+}
+
+// newKubeResource creates a new TeleportKubernetesResource
+// the function only supports structs or pointer to struct implementations of the TeleportKubernetesResource interface
+func newKubeResource[T types.ResourceWithOrigin, K TeleportKubernetesResource[T]]() (K, error) {
+	// We create a new instance of K.
+	var resource K
+	// We take the type of K
+	interfaceType := reflect.TypeOf(resource)
+	// If K is not a pointer we don't need to do anything
+	// If K is a pointer, new(K) is only initializing a nil pointer, we need to manually initialize its destination
+	if interfaceType.Kind() == reflect.Ptr {
+		if interfaceType.Elem().Kind() != reflect.Struct {
```

Comment:
IIRC, I think it's not possible to satisfy this part of the code. To satisfy an interface with methods I think go only allows a structure or a single pointer to a structure.

This return is not necessary.

Can we add some tests to this function to make sure it's properly covered?

Reply:
done